### PR TITLE
Massive refactor

### PR DESCRIPTION
**Pull request by kleewho**
The main idea behind this change is as follows:

* create structure in endpoints.go that will hold pubnub and ctx, the common things across all *Opts structures
* make this endpointOpts a first anonymous field all other *Opts structures, so all methods implemented for endpointOpts will work for them as well
* create a factory methods for all *Opts, so it's less error prone to create them. pubnub and ctx are basically required. All those factory methods should be used in tests and in the library code as well instead of directly creating structs
* move all methods implemented in all endpoints in exactly the same way to endpoints.go
* move all methods implemented in almost all endpoints in exactly the same way to endpoints.go. This provides the common default implementation, but can and is overiden in some places

### CODE REVIEW

File: history_delete_request_test.go [hunk 1]
```diff
@@ -12,14 +12,12 @@ import (
...
 
-	opts := &historyDeleteOpts{
-		Channel:  "ch",
-		SetStart: true,
-		SetEnd:   true,
-		Start:    int64(123),
-		End:      int64(456),
-		pubnub:   pubnub,
-	}
+	opts := newHistoryDeleteOpts(pubnub, pubnub.ctx)
```
Comment (Xavrax) [inline history_delete_request_test.go:15]: same here

File: publish_request_test.go [hunk 2]
```diff
@@ -414,19 +405,19 @@ func TestPublishStore(t *testing.T) {
...
+	opts := newPublishOpts(pn, pn.ctx)
+	opts.Channel = "ch"
+	opts.Message = "hey"
 
 	path, err := opts.buildPath()
 	assert.Nil(err)
 
 	assert.Equal(
-		"/publish/pub_key/sub_key/0/ch/0/%22%2Bc52pEK3TCTpuEjEFzukRw%3D%3D%22", path)
+		"/publish/demo/demo/0/ch/0/%22MnwzPGdVgz2osQCIQJviGg%3D%3D%22", path)
```
Comment (Xavrax) [inline publish_request_test.go:420]: Is this an intentional change? I don't get where the implementation changes ;/.

**Review by Xavrax — APPROVED** [review]
LGTM

**Review by mateusz-dahlke — APPROVED** [review]
image

File: publish_request_test.go (hunk 2 above)
Comment (kleewho) [reply publish_request_test.go:420]: Yes, this was an intentional change. The default `Serialize: true` was added by `newPublishOpts` constructing method but it was not used in the test previously, therefore the output should change

File: grant_request_test.go [hunk 3]
```diff
@@ -82,17 +82,15 @@ func TestNewGrantObjectsPremsBuilder(t *testing.T) {
...
-	}
+	opts := newGrantOpts(pubnub, pubnub.ctx)
+	opts.AuthKeys = []string{"my-auth-key"}
+	opts.Channels = []string{"ch"}
+	opts.ChannelGroups = []string{"cg"}
+	opts.Read = true
+	opts.Write = true
+	opts.Manage = true
+	opts.TTL = 5000
+	opts.setTTL = true
```
[the remark this answers is not in the document]
Comment (kleewho) [reply grant_request_test.go:93]: I'll try to do that

File: history_delete_request_test.go (hunk 1 above)
Comment (kleewho) [reply history_delete_request_test.go]: Sure

File: endpoints.go [hunk 4]
```diff
@@ -15,9 +16,14 @@ import (
...
-type endpointOpts interface {
+type endpointOpts struct {
+	pubnub *PubNub
+	ctx    Context
+}
+
+type endpoint interface {
 	jobQueue() chan *JobQItem
-	config() Config
+	config() *Config
```
[the remark this answers is not in the document]
Comment (kleewho) [reply endpoints.go:26]: It took me a while to understand why exactly, but I think that it's not necessarily the same. If we go the pointer to the mutex route keeping all other fields still values and not pointers we will effective have just one mutex instance possibly guarding different set of values. 

Also a little bit different case. Currently this lock makes sure that two fields always change together. If we decide to return config by value nothing guarantees that the copying won't happen exactly in between changing first field but before changing the second field, which renders the locking useless. 

Therefore I keep pointer to the Config.

**Pull request merged**